- **토치텍스트(Torchtext)** : PyTorch에서 텍스트에 대한 여러 추상화 기능을 제공하는 자연어 처리 라이브러리
- 토치텍스트의 기능
    - 파일 로드(Flie Loading) : 다양한 포맷의 코퍼스 로드
    - 토큰화(Tokenization) : 단어 단위로 문장 분리
    - 단어 집합(Vocab) : 단어 집합 생성
    - 정수 인코딩(Integer encoding) : 전체 코퍼스의 단어들을 각각의 고유한 정수로 맵핑
    - 단어 벡터(Word Vector) : 단어 집합의 단어들에 고유한 임베딩 벡터를 만든다. 랜덤값으로 초기화한 값이거나 사전 훈련된 임베딩 벡터를 로드할 수도 있다.
    - 배치화(Batching) : 훈련 샘플들의 배치를 만든다. 이 과정에서 패딩(Padding) 작업도 이루어진다.


# 훈련 데이터, 테스트 데이터 분리

In [25]:
import urllib.request
import pandas as pd

In [26]:
# IMDB 리뷰 데이터 다운로드
urllib.request.urlretrieve("https://raw.githubusercontent.com/LawrenceDuan/IMDb-Review-Analysis/master/IMDb_Reviews.csv", filename="IMDb_Reviews.csv")

('IMDb_Reviews.csv', <http.client.HTTPMessage at 0x18a822d64f0>)

In [27]:
df = pd.read_csv('IMDb_Reviews.csv', encoding='latin1')
df.head()

review  sentiment
0  My family and I normally do not watch local mo...          1
1  Believe it or not, this was at one time the wo...          0
2  After some internet surfing, I found the "Home...          0
3  One of the most unheralded great works of anim...          1
4  It was the Sixties, and anyone with long hair ...          0

In [28]:
print('전체 샘플의 개수 : {}'.format(len(df)))

전체 샘플의 개수 : 50000


In [29]:
train_df = df[:25000]
test_df = df[25000:]

In [30]:
train_df.to_csv("train_data.csv", index=False)
test_df.to_csv("test_data.csv", index=False)

# 필드 정의 (torchtext.data)

- 필드를 통해 앞으로 어떤 전처리를 할 것인지 정의한다.

In [31]:
from torchtext.legacy import data # torchtext.data 임포트

# 필드 정의
TEXT = data.Field(sequential=True,
                  use_vocab=True,
                  tokenize=str.split,
                  lower=True,
                  batch_first=True,
                  fix_length=20)

LABEL = data.Field(sequential=False,
                   use_vocab=False,
                   batch_first=False,
                   is_target=True)

- TEXT : 실제 테스트를 위한 객체
- LABEL : 레이블 데이터를 위한 객체

- 각 인자에 대한 설명
    - sequential : 시퀀스 데이터 여부. (True가 기본값)
    -use_vocab : 단어 집합을 만들 것인지 여부. (True가 기본값)
    -tokenize : 어떤 토큰화 함수를 사용할 것인지 지정. (string.split이 기본값)
    - lower : 영어 데이터를 전부 소문자화한다. (False가 기본값)
    - batch_first : 미니 배치 차원을 맨 앞으로 하여 데이터를 불러올 것인지 여부. (False가 기본값)
    - is_target : 레이블 데이터 여부. (False가 기본값)
    - fix_length : 최대 허용 길이. 이 길이에 맞춰서 패딩 작업(Padding)이 진행된다.

# 데이터셋 만들기

In [32]:
from torchtext.legacy.data import TabularDataset

- TabularDataset은 데이터를 불러오면 필드에서 정의한 토큰화 방법으로 토큰화를 수행하며 소문자화 같은 기본적이 전처리가 함께 이루어진다.

In [33]:
train_data, test_data = TabularDataset.splits(
        path='.', train='train_data.csv', test='test_data.csv', format='csv',
        fields=[('text', TEXT), ('label', LABEL)], skip_header=True)

- path : 파일이 위치한 경로.
- format : 데이터의 포맷.
- fields : 위에서 정의한 필드를 지정. 첫번째 원소는 데이터 셋 내에서 해당 필드를 호칭할 이름, 두번째 원소는 지정할 필드.
- skip_header : 데이터의 첫번째 줄은 무시.

In [34]:
print('훈련 샘플의 개수 : {}'.format(len(train_data)))
print('테스트 샘플의 개수 : {}'.format(len(test_data)))

훈련 샘플의 개수 : 25000
테스트 샘플의 개수 : 25000


In [35]:
print(vars(train_data[0]))

{'text': ['my', 'family', 'and', 'i', 'normally', 'do', 'not', 'watch', 'local', 'movies', 'for', 'the', 'simple', 'reason', 'that', 'they', 'are', 'poorly', 'made,', 'they', 'lack', 'the', 'depth,', 'and', 'just', 'not', 'worth', 'our', 'time.<br', '/><br', '/>the', 'trailer', 'of', '"nasaan', 'ka', 'man"', 'caught', 'my', 'attention,', 'my', 'daughter', 'in', "law's", 'and', "daughter's", 'so', 'we', 'took', 'time', 'out', 'to', 'watch', 'it', 'this', 'afternoon.', 'the', 'movie', 'exceeded', 'our', 'expectations.', 'the', 'cinematography', 'was', 'very', 'good,', 'the', 'story', 'beautiful', 'and', 'the', 'acting', 'awesome.', 'jericho', 'rosales', 'was', 'really', 'very', 'good,', "so's", 'claudine', 'barretto.', 'the', 'fact', 'that', 'i', 'despised', 'diether', 'ocampo', 'proves', 'he', 'was', 'effective', 'at', 'his', 'role.', 'i', 'have', 'never', 'been', 'this', 'touched,', 'moved', 'and', 'affected', 'by', 'a', 'local', 'movie', 'before.', 'imagine', 'a', 'cynic', 'like', 'me

In [36]:
# 필드 구성 확인.
print(train_data.fields.items())

dict_items([('text', <torchtext.legacy.data.field.Field object at 0x0000018A57A3C5E0>), ('label', <torchtext.legacy.data.field.Field object at 0x0000018A57A3CFA0>)])


# Vocabulary 만들기

In [37]:
TEXT.build_vocab(train_data, min_freq=10, max_size=10000)

In [38]:
print('단어 집합의 크기 : {}'.format(len(TEXT.vocab)))

단어 집합의 크기 : 10002


In [ ]:
print(TEXT.vocab.stoi)

- 단어 집합의 크기를 10000개로 제한했지만 실제로는 0~10001번까지 총 10002개인 이유는 토치텍스트에서 임의로 특별 토큰 **<unk\>** 와 **<pad\>** 를 추가했기 때문이다.
- <unk\> : 집합에 없는 단어 표현
- <pad\> : 길이 맞추는 패딩 작업

# 토치텍스트의 데이터로더 만들기

In [40]:
from torchtext.legacy.data import Iterator

In [41]:
batch_size = 5

In [42]:
train_loader = Iterator(dataset=train_data, batch_size = batch_size)
test_loader = Iterator(dataset=test_data, batch_size = batch_size)

In [43]:
print('훈련 데이터의 미니 배치 수 : {}'.format(len(train_loader)))
print('테스트 데이터의 미니 배치 수 : {}'.format(len(test_loader)))

훈련 데이터의 미니 배치 수 : 5000
테스트 데이터의 미니 배치 수 : 5000


In [44]:
batch = next(iter(train_loader)) # 첫번째 미니배치

In [45]:
print(type(batch))

<class 'torchtext.legacy.data.batch.Batch'>


- 일반적 데이터로더 : 미니 배치를 텐서로 가져온다.
- 토치텍스트 데이터로더 : torchtext.data.batch.Batch 객체를 가져온다.

In [46]:
print(batch.text)

tensor([[1959,    0,    0,    0,   41,   21,  677,  556,    4,   21,  397,    0,
         2935,    8,    3,  426,    0,  848,    0,    0,    0,  436,    6,  328,
          101,  143,    2, 2397, 9248,    0,   97,    0,  607,  368, 6377,   17,
            3,  161,    0,    0, 3226,    4,    0,    6,    3,    0,   27,    7,
          552,   10],
        [  10,    7,    3,   20,   17,    3,  422, 7483,   18,   49,  953, 5003,
           11,  133,   99,   26,   71,  685,   17,   47,   29,  225,  278,   51,
            0,    0,   11,  178,    0,    2,   77,    7,  105,    0, 1630, 1691,
            4,   52,   23,  105,  100, 1826, 3283,  454,    8,    2,   77,    6,
           90,   11],
        [   9,  217,    4,   14,   49, 3413,   31,    2,  118,   11,   14,   55,
           36,   84,  487,   20,    5,    0,  102, 2881,    8,    2, 2165,  612,
            9,   14,   38,    8,  113,   17, 8186,    0,  166,    4,    0,  419,
            9,  217,   60,   20,    0,   16,    2,   84,    0,   

- 각 샘플은 더 이상 단어 시퀀스가 아니라 정수 시퀀스임을 볼 수 있다.
- 중간에는 숫자 0이 존재하는데 이는 <unk\> 토큰의 번호로 단어 집합에 포함되지 못한 단어들은 <unk\>라는 토큰으로 변환되었다.

# <pad\> 토큰이 사용되는 경우

In [47]:
batch = next(iter(train_loader)) # 첫번째 미니배치
print(batch.text[0]) # 첫번째 미니배치 중 첫번째 샘플

tensor([ 189,   48,   30,    5,    2,  643,  350,   29,    2, 2831,   13, 6797,
           2, 5482,  594,    3, 6097,   16,    3,    0,  594,  193,    2,  643,
         822,   51,    0,    3,    0,    0,    4,    2,    0, 8416,    0,  867,
           6,    3,  985,  189,    5,    0,  103,   13,  793,   30,    5,    2,
          78, 3770])
